application/src/main/java/org/opentripplanner/transit/model/timetable/RealTimeTripTimesBuilder.java

```diff
@@ -0,0 +1,371 @@
+
+  public RealTimeTripTimesBuilder withCanceled(int stop) {
+    return withStopRealTimeState(stop, StopRealTimeState.CANCELLED);
+  }
+
+  public RealTimeTripTimesBuilder withNoData(int stop) {
+    return withStopRealTimeState(stop, StopRealTimeState.NO_DATA);
+  }
+
+  public RealTimeTripTimesBuilder withInaccuratePredictions(int stop) {
+    return withStopRealTimeState(stop, StopRealTimeState.INACCURATE_PREDICTIONS);
+  }
+
+  private RealTimeTripTimesBuilder withStopRealTimeState(int stop, StopRealTimeState state) {
+    this.stopRealTimeStates[stop] = state;
+    return this;
+  }
+
+  @Nullable
+  public I18NString tripHeadsign() {
+    if (tripHeadsign == null) {
+      return scheduledTripTimes.getTripHeadsign();
+    }
+    return tripHeadsign;
+  }
+
+  public RealTimeTripTimesBuilder withTripHeadsign(I18NString headsign) {
+    tripHeadsign = headsign;
+    return this;
+  }
+
+  public @Nullable I18NString[] stopHeadsigns() {
+    var result = scheduledTripTimes.copyHeadsigns(() ->
+      new I18NString[scheduledTripTimes.getNumStops()]
+    );
+    for (var i = 0; i < result.length; i++) {
+      if (stopHeadsigns[i] != null) {
+        result[i] = stopHeadsigns[i];
+      }
+    }
+    return result;
+  }
+
+  public RealTimeTripTimesBuilder withStopHeadsign(int stop, I18NString headsign) {
+    stopHeadsigns[stop] = headsign;
+    return this;
+  }
+
+  public OccupancyStatus[] occupancyStatus() {
+    return occupancyStatus.clone();
+  }
+
+  public RealTimeTripTimesBuilder withOccupancyStatus(int stop, OccupancyStatus occupancyStatus) {
+    this.occupancyStatus[stop] = occupancyStatus;
+    return this;
+  }
+
+  public Accessibility wheelchairAccessibility() {
+    if (wheelchairAccessibility == null) {
+      return scheduledTripTimes.getWheelchairAccessibility();
+    }
+    return wheelchairAccessibility;
+  }
+
+  public RealTimeTripTimesBuilder withWheelchairAccessibility(
+    Accessibility wheelchairAccessibility
+  ) {
+    this.wheelchairAccessibility = wheelchairAccessibility;
+    return this;
+  }
+
+  public RealTimeTripTimesBuilder withServiceCode(int serviceCode) {
+    this.scheduledTripTimes = scheduledTripTimes
+      .copyOfNoDuplication()
+      .withServiceCode(serviceCode)
+      .build();
+    return this;
+  }
+
+  /**
+   * Note: This method only applies for GTFS, not SIRI!
+   * This method interpolates the times for SKIPPED stops in between regular stops since GTFS-RT
+   * does not require arrival and departure times for these stops. This method ensures the internal
+   * time representations in OTP for SKIPPED stops are between the regular stop times immediately
+   * before and after the cancellation in GTFS-RT. This is to meet the OTP requirement that stop
+   * times should be increasing and to support the trip search flag `includeRealtimeCancellations`.
+   * Terminal stop cancellations can be handled by backward and forward propagations, and are
+   * outside the scope of this method.
+   *
```

Comment:
Great comment! Thank you.

Reply:
This is just moved verbatim from the old class and will be worked on next, to extract the interpolation out of the builder.